python/cuml/common/memory_utils.py

```diff
@@ -255,40 +257,57 @@ def _rmm_cupy6_array_like(ary, order):
     return arr
 
 
-def _strides_to_order(strides, dtype):
+def _strides_to_order(strides, shape, dtype):
     # cuda array interface specification
     if strides is None:
         return 'C'
-    if strides[0] == dtype.itemsize or len(strides) == 1:
+
+    itemsize = cp.dtype(dtype).itemsize
+    shape = list(shape)
+    strides = list(strides)
+
+    if strides[0] == itemsize and \
+        all(map(lambda i: strides[i + 1] == strides[i] * shape[i],
+                range(len(strides) - 1))):
         return 'F'
-    elif strides[1] == dtype.itemsize:
+
+    shape = shape[::-1]
+    strides = strides[::-1]
+
+    if strides[0] == itemsize and \
+        all(map(lambda i: strides[i + 1] == strides[i] * shape[i],
+                range(len(strides) - 1))):
         return 'C'
     else:
-        raise ValueError("Invalid strides value for dtype")
+        debug('Uncontiguous array, will perform a copy')
+        return None
 
 
 def _order_to_strides(order, shape, dtype):
     itemsize = cp.dtype(dtype).itemsize
-    if isinstance(shape, int):
+    if isinstance(shape, int) or len(shape) == 1:
         return (itemsize, )
 
     elif len(shape) == 0:
         return None
 
-    elif len(shape) == 1:
-        return (itemsize, )
+    strides = [itemsize]
 
-    elif order == 'C':
-        dim_minor = shape[1] * itemsize
-        return (dim_minor, itemsize)
+    if order == 'F':
+        for dim_size in shape[:-1]:
+            strides.append(dim_size * strides[-1])
 
-    elif order == 'F':
-        dim_minor = shape[0] * itemsize
-        return (itemsize, dim_minor)
+    elif order == 'C':
+        shape = shape[::-1]
+        for dim_size in shape[:-1]:
+            strides.append(dim_size * strides[-1])
+        strides = strides[::-1]
```

Comment:
We could consider to use a combination of `itertools.accumulate()` and `operator.mul()` to compute the strides a bit more succinctly:

```python
from itertools import accumulate
from operator import mul


f_strides = list(accumulate(shape[:-1], func=mul, initial=item_size))
c_strides = list(accumulate(shape[:0:-1], func=mul, initial=item_size))[::-1]
```

Reply:
Looks like loops are the fastest, however I was able to apply a micro optimization that is a bit faster (see below):
```
Loops:              100.00%
Loops (alt):        94.73%
With accumulate:    132.87%
```
I ran this a few times and results appear pretty consistent.

<details>
<summary>Benchmark code</summary>

```python
from itertools import accumulate
from operator import mul


def compute_strides(shape, item_size):
    tuple(accumulate(shape[:-1], func=mul, initial=item_size))
    tuple(accumulate(shape[:0:-1], func=mul, initial=item_size))[::-1]


def compute_strides_loops(shape, item_size):

    strides = [item_size]
    for dim_size in shape[:-1]:
        strides.append(dim_size * strides[-1])
    tuple(strides)

    strides = [item_size]
    shape = shape[::-1]
    for dim_size in shape[:-1]:
        strides.append(dim_size * strides[-1])
    tuple(strides[::-1])


def compute_strides_loops_alternative(shape, item_size):

    strides = [item_size]
    for dim_size in shape[:-1]:
        strides.append(dim_size * strides[-1])
    tuple(strides)

    strides = [item_size]
    for dim_size in shape[:0:-1]:
        strides.append(dim_size * strides[-1])
    tuple(strides[::-1])


if __name__ == "__main__":
    from timeit import timeit

    result_w_loops = timeit(
        "compute_strides((2, 3), 8)",
        setup="from benchmark_strides import compute_strides_loops as compute_strides",
    )

    result_w_loops_alt = timeit(
        "compute_strides((2, 3), 8)",
        setup="from benchmark_strides import compute_strides_loops_alternative as compute_strides",
    )

    result_w_accumulate = timeit(
        "compute_strides((2, 3), 8)",
        setup="from benchmark_strides import compute_strides",
    )

    print(f"Loops:              {result_w_loops / result_w_loops:.2%}")
    print(f"Loops (alt):        {result_w_loops_alt / result_w_loops:.2%}")
    print(f"With accumulate:    {result_w_accumulate / result_w_loops:.2%}")

```
</details>

Micro-optimization:
```python
    # F-order
    strides = [item_size]
    for dim_size in shape[:-1]:
        strides.append(dim_size * strides[-1])
    tuple(strides)

   # C-order
    strides = [item_size]
    for dim_size in shape[:0:-1]:
        strides.append(dim_size * strides[-1])
    tuple(strides[::-1])
```